### PR TITLE
[Flow] Fold flow reshape with mismatching dyn dims

### PR DESCRIPTION
**Pull request by IanWood1**
`flow.tensor.reshape` is foldable when there is a single mismatching dynamic dimension (unequal SSA values) because they must be the same.

Cleans up pattern found in migx_bert-large-uncased

### CODE REVIEW

File: compiler/src/iree/compiler/Dialect/Flow/IR/FlowOpFolders.cpp [hunk 1]
```diff
@@ -770,7 +769,7 @@ static bool compareShapesEqual(ShapedType lhsType, ValueRange lhsDynamicDims,
       }
     }
   }
-  return true;
+  return numNonmatchingSSADims < 2;
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Flow/IR/FlowOpFolders.cpp]: nit: `<= 1` (just reads easier - "0 or 1" vs "less than 2")

File: compiler/src/iree/compiler/Dialect/Flow/IR/test/tensor_folding.mlir [hunk 2]
```diff
@@ -112,7 +112,7 @@ util.func public @reshapeNoOpDynamic(%arg0: tensor<4x?xf32>, %dim: index) -> ten
 
 // CHECK-LABEL: @reshapeDynamicDifferent
 util.func public @reshapeDynamicDifferent(%arg0: tensor<4x?xf32>, %dim0: index, %dim1: index) -> tensor<4x?xf32> {
-  // CHECK-NEXT: flow.tensor.reshape %arg0
+  // CHECK-NEXT: util.return %arg0 : tensor<4x?xf32>
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Flow/IR/test/tensor_folding.mlir:115]: maybe add a new test with two values instead to ensure it's preserved

Comment (IanWood1) [reply compiler/src/iree/compiler/Dialect/Flow/IR/test/tensor_folding.mlir:115]: Good point + checks that it doesnt get completely folded

**Review by MaheshRavishankar — COMMENTED** [review]
Am I missing something here. The test added here does not seem to be an example of what is described in the PR description?

**Comment by IanWood1** [comment]
Are you looking at `foldReshapeChain`? The diff is a bit confusing. `flattenReshapeChain` was the old test to combine multiple reshapes, but because it only has 1 dynamic dim it got completely folded away. `foldReshapeChain` is the new test (which uses the old mlir input), but the difference is that it checks that all reshape ops get folded. But I could be wrong

**Comment by MaheshRavishankar** [comment]
> Are you looking at `foldReshapeChain`? The diff is a bit confusing. `flattenReshapeChain` was the old test to combine multiple reshapes, but because it only has 1 dynamic dim it got completely folded away. `foldReshapeChain` is the new test (which uses the old mlir input), but the difference is that it checks that all reshape ops get folded. But I could be wrong

Ah interesting. I did miss that, I see now. I dont know how this change made `flatterReshapeChain` work.

**Review by MaheshRavishankar — APPROVED** [review]

**Pull request merged**